src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1360,6 +1425,432 @@ private Object writeReplace() {
+            return size;
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return false;
+        }
+
+        @Override
+        public StableValue<V> get(Object key) {
+            int i = probe(key);
+            if (i >= 0) {
+                return value(i);
+            } else {
+                return null;
+            }
+        }
+
+        final class StableMapIterator implements Iterator<Map.Entry<K, StableValue<V>>> {
+
+            private int remaining;
+            private int idx;
+
+            StableMapIterator() {
+                remaining = size;
+                // pick an even starting index in the [0, keys.length)
+                // range randomly based on SALT32L
+                idx = (int) ((SALT32L * (keys.length)) >>> 32);
+            }
+
+            @Override
+            public boolean hasNext() {
+                return remaining > 0;
+            }
+
+            private int nextIndex() {
+                int idx = this.idx;
+                if (REVERSE) {
+                    if ((++idx) >= keys.length) {
+                        idx = 0;
+                    }
+                } else {
+                    if ((--idx) < 0) {
+                        idx = keys.length - 1;
+                    }
+                }
+                return this.idx = idx;
+            }
+
+            @Override
+            public Map.Entry<K, StableValue<V>> next() {
+                if (remaining > 0) {
+                    int idx;
+                    while (keys[idx = nextIndex()] == null) {}
+                    Map.Entry<K, StableValue<V>> e = new KeyValueHolder<>(keys[idx], value(idx));
+                    remaining--;
+                    return e;
+                } else {
+                    throw new NoSuchElementException();
+                }
+            }
+        }
+
+        @Override
+        public Set<Map.Entry<K, StableValue<V>>> entrySet() {
+            return new AbstractSet<>() {
+                @Override
+                public int size() {
+                    return StableMap.this.size;
+                }
+
+                @Override
+                public Iterator<Map.Entry<K, StableValue<V>>> iterator() {
+                    return new StableMapIterator();
+                }
+            };
+        }
+
+        @Override
+        public V computeIfUnset(K key, Function<? super K, ? extends V> mapper) {
+           StableValueElement<V> element = (StableValueElement<V>) get(key);
+           if (element == null) {
+               throw noKey(key);
+           }
+            return element.computeIfUnset(key, mapper);
+        }
+
+        static <K, V> Map<K, StableValue<V>> create(K[] keys) {
+            return new StableMap<>(keys);
+        }
+
+    }
+
+    static final class StableEnumMap<K extends Enum<K>, V>
+            extends AbstractImmutableMap<K, StableValue<V>>
+            implements Map<K, StableValue<V>>, HasComputeIfUnset<K, V> {
```

Comment:
While that is true, no other immutable collection implements a navigable map. The way the API is currently wired, it always returns a `Map`. If we go down this route, it would incidentally return a `NaviableMap` if presented with an `EnumMap` or, we could have a separate factory for enums that states it returns a `NavigableMap`. I think creating all the required views would increase complexity significantly and I am not sure it would be used that much. That said, let us keep this open for the future.

Reply:
Fair enough, `Collections` APIs like `unmodifiableSortedMap` explicitly avoid implementing too many interfaces.